core.js

```diff
@@ -0,0 +1,131 @@
+      var params = {type: type};
+
+      // Ensure that we have a URL.
+      if (!options.url) {
+          options.url = result(model, 'url') || urlError();
+      }
+
+      // Ensure that we have the appropriate request data.
+      if (options.data == null && model && (method === 'create' || method === 'update' || method === 'patch')) {
+          params.json = options.attrs || model.toJSON(options);
+      }
+
+      // If passed a data param, we add it to the URL or body depending on request type
+      if (options.data && type === 'GET') {
+          // make sure we've got a '?'
+          options.url += includes(options.url, '?') ? '&' : '?';
+          options.url += qs.stringify(options.data);
+      }
+
+      // For older servers, emulate JSON by encoding the request into an HTML-form.
+      if (options.emulateJSON) {
+          headers['Content-Type'] = 'application/x-www-form-urlencoded';
+          params.body = params.json ? {model: params.json} : {};
+          delete params.json;
+      }
+
+      // For older servers, emulate HTTP by mimicking the HTTP method with `_method`
+      // And an `X-HTTP-Method-Override` header.
+      if (options.emulateHTTP && (type === 'PUT' || type === 'DELETE' || type === 'PATCH')) {
+          params.type = 'POST';
+          if (options.emulateJSON) params.body._method = type;
+          headers['X-HTTP-Method-Override'] = type;
+      }
+
+      // When emulating JSON, we turn the body into a querystring.
+      // We do this later to let the emulateHTTP run its course.
+      if (options.emulateJSON) {
+          params.body = qs.stringify(params.body);
+      }
+
+      // Start setting ajaxConfig options (headers, xhrFields).
+      var ajaxConfig = (result(model, 'ajaxConfig') || {});
+
+      // Combine generated headers with user's headers.
+      if (ajaxConfig.headers) {
+          assign(headers, ajaxConfig.headers);
+      }
+      params.headers = headers;
+
+      //Set XDR for cross domain in IE8/9
+      if (ajaxConfig.useXDR) {
+          params.useXDR = true;
+      }
+
+      // Set raw xhr options.
+      if (ajaxConfig.xhrFields) {
+          var beforeSend = ajaxConfig.beforeSend;
+          params.beforeSend = function (req) {
+              for (var key in ajaxConfig.xhrFields) {
+                  req[key] = ajaxConfig.xhrFields[key];
+              }
+              if (beforeSend) return beforeSend.apply(this, arguments);
+          };
+          params.xhrFields = ajaxConfig.xhrFields;
+      } else {
+          params.beforeSend = ajaxConfig.beforeSend;
+      }
+
+      // Turn a jQuery.ajax formatted request into xhr compatible
+      params.method = params.type;
+
+      var ajaxSettings = assign(params, options);
+
+      // Make the request. The callback executes functions that are compatible
+      // With jQuery.ajax's syntax.
+      var request = options.xhr = options.xhrImplementation(ajaxSettings, function (err, resp, body) {
```

Comment:
I never understood how overwriting options.xhr with itself benefited this code, but now it's probably going to be harmful, as `request` doesn't return itself.